### ISSUE TITLE
Fatal Error parsing invalid cert aborts unit tests

### ISSUE THREAD

**Issue by falko-strenzke** [zmap/zlint#324]
I tried to use the certificate below as test-data in a unit test. The certificate is invalid since it features a GeneralString in the subject:organizationIdentifier. When trying to parse the certificate this leads to

    FATA[0000] unable to parse PEM                          
    exit status 1

During unit tests, this leads to the abortion of the tests due to a panic. I think the unit test framework should at least catch any errors of this kind and not simply abort. Even better if the it would be possible to parse such invalid certificates.

-----BEGIN CERTIFICATE-----
MIIEyTCCA7GgAwIBAgINBnUDUrQShFEl9jfPmDANBgkqhkiG9w0BAQsFADBBMRUwEwYDVQQDDAxUZXN0IFN1YkNBIDIxDTALBgNVBAsMBFRlc3QxDDAKBgNVBAoMA01URzELMAkGA1UEBhMCREUwHhcNMTkxMTA0MDYzNTM2WhcNMjAxMTA0MDYzNTM2WjCBojEMMAoGA1UECgwDTVRHMRIwEAYDVQQHDAlEYXJtc3RhZHQxDzANBgNVBAgMBkhlc3NlbjELMAkGA1UEBhMCREUxHTAbBgNVBA8MFFByaXZhdGUgT3JnYW5pemF0aW9uMRgwFgYDVQRhGw9WQVRERS0xMjM0NTY3ODkxEjAQBgNVBAUTCTEyMzQ1Njc4OTETMBEGCysGAQQBgjc8AgEDDAJERTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANjJFjxxhlF2HKfm9gFWp4DMwUS5Qhb9Jomq2O425As5PCe9VCFNmV1X8a6wmADuWPPH3UD0BleveWnbtsj1IpU1Eo6uPwqeZQGJIKRuz6wgx9AZcbpXT5zgmf5VDDoVFn7sYvSc4a65FkfE2Ontamf0YuMDfpbZRS7nBzfojQFzKiB0OGvyC4icBMwGm+y11qwpuOImTzRljj3rAEboazF3o+8S8TyY712aXPQno4c3MSM1EqtZO+ILSBUy0ULzVWMFybI/WsnlqkpIpHQSFDX7V5ck8uiLLPrb/WnuE9EQXj4c2B8m8WlpUlAP9fMtqeddOqWSnxx4FAxhyUp9DPUCAwEAAaOCAVwwggFYMB8GA1UdIwQYMBaAFJZgP5yaf4gl4AHNZwhX0IYm8PFKMB0GA1UdDgQWBBTgmXRqGmcO54swL6B3XM95sawpYTAOBgNVHQ8BAf8EBAMCBaAwDAYDVR0TAQH/BAIwADAaBgNVHREEEzARgg93d3cuZXhhbXBsZS5jb20wYgYIKwYBBQUHAQEEVjBUMCgGCCsGAQUFBzAChhxodHRwOi8vY2EuZXhhbXBsZS5jb20vY2EuY3J0MCgGCCsGAQUFBzABhhxodHRwOi8vb2NzcC5leGFtcGxlLmNvbS9vY3NwMFkGA1UdIARSMFAwBwYFZ4EMAQEwRQYKKwYBBAG9Rw0YATA3MDUGCCsGAQUFBwIBFilodHRwOi8vd3d3LnRlbGVzZWMuZGUvc2VydmVycGFzcy9jcHMuaHRtbDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDQYJKoZIhvcNAQELBQADggEBAFRO6OOXlSsTAKv4Hp2jvIzMQIwwTZJMkHs4qshiFVOrLSXe3dDxxBjgn87rbXL0X6rqHysnSzhlhbW6bUg2gAAcyXf4DBu/TEwYUhlLkzSFFHM9tioHWZmL1sgedl2x4vwnZbMpqRlQTINwWbV0Ey5TfqGw84LoC0SXCzRoV3ta1tsZDIXS5UBuEZXMJ2ElvjqYbEnJhHjUqWRuCtbVlRHJ05EqigDLzv2Qm0Q5MluQq/Qc4SK1WZQ+PGOv2yU/VXdItYirD4LYBAsI2o2Abq5o/tjtjQeynj3NxqrF1Qd0/SHjQXE+ZMU+M7glP9e3jjwd9liavUKsF8O6wLha73g=-----END CERTIFICATE-----

**Comment by zakird** [comment]
We agree, this is not expected behavior. If ZLint is unable to parse a certificate that should result with the lint failing with a `FATAL` not for the library to panic.

**Comment by zakird** [comment]
Actually, scratch that last comment. This certificate does not cause the library to panic. When the certificate is in the form you presented:
```
-----BEGIN CERTIFICATE-----
MIIEyTCCA7GgAwIBAgINBnUDUrQShFEl9jfPmDANBgkqhkiG9w0BAQsFADBBMRUwEwYDVQQDDAxUZXN0IFN1YkNBIDIxDTALBgNVBAsMBFRlc3QxDDAKBgNVBAoMA01URzELMAkGA1UEBhMCREUwHhcNMTkxMTA0MDYzNTM2WhcNMjAxMTA0MDYzNTM2WjCBojEMMAoGA1UECgwDTVRHMRIwEAYDVQQHDAlEYXJtc3RhZHQxDzANBgNVBAgMBkhlc3NlbjELMAkGA1UEBhMCREUxHTAbBgNVBA8MFFByaXZhdGUgT3JnYW5pemF0aW9uMRgwFgYDVQRhGw9WQVRERS0xMjM0NTY3ODkxEjAQBgNVBAUTCTEyMzQ1Njc4OTETMBEGCysGAQQBgjc8AgEDDAJERTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANjJFjxxhlF2HKfm9gFWp4DMwUS5Qhb9Jomq2O425As5PCe9VCFNmV1X8a6wmADuWPPH3UD0BleveWnbtsj1IpU1Eo6uPwqeZQGJIKRuz6wgx9AZcbpXT5zgmf5VDDoVFn7sYvSc4a65FkfE2Ontamf0YuMDfpbZRS7nBzfojQFzKiB0OGvyC4icBMwGm+y11qwpuOImTzRljj3rAEboazF3o+8S8TyY712aXPQno4c3MSM1EqtZO+ILSBUy0ULzVWMFybI/WsnlqkpIpHQSFDX7V5ck8uiLLPrb/WnuE9EQXj4c2B8m8WlpUlAP9fMtqeddOqWSnxx4FAxhyUp9DPUCAwEAAaOCAVwwggFYMB8GA1UdIwQYMBaAFJZgP5yaf4gl4AHNZwhX0IYm8PFKMB0GA1UdDgQWBBTgmXRqGmcO54swL6B3XM95sawpYTAOBgNVHQ8BAf8EBAMCBaAwDAYDVR0TAQH/BAIwADAaBgNVHREEEzARgg93d3cuZXhhbXBsZS5jb20wYgYIKwYBBQUHAQEEVjBUMCgGCCsGAQUFBzAChhxodHRwOi8vY2EuZXhhbXBsZS5jb20vY2EuY3J0MCgGCCsGAQUFBzABhhxodHRwOi8vb2NzcC5leGFtcGxlLmNvbS9vY3NwMFkGA1UdIARSMFAwBwYFZ4EMAQEwRQYKKwYBBAG9Rw0YATA3MDUGCCsGAQUFBwIBFilodHRwOi8vd3d3LnRlbGVzZWMuZGUvc2VydmVycGFzcy9jcHMuaHRtbDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDQYJKoZIhvcNAQELBQADggEBAFRO6OOXlSsTAKv4Hp2jvIzMQIwwTZJMkHs4qshiFVOrLSXe3dDxxBjgn87rbXL0X6rqHysnSzhlhbW6bUg2gAAcyXf4DBu/TEwYUhlLkzSFFHM9tioHWZmL1sgedl2x4vwnZbMpqRlQTINwWbV0Ey5TfqGw84LoC0SXCzRoV3ta1tsZDIXS5UBuEZXMJ2ElvjqYbEnJhHjUqWRuCtbVlRHJ05EqigDLzv2Qm0Q5MluQq/Qc4SK1WZQ+PGOv2yU/VXdItYirD4LYBAsI2o2Abq5o/tjtjQeynj3NxqrF1Qd0/SHjQXE+ZMU+M7glP9e3jjwd9liavUKsF8O6wLha73g=-----END CERTIFICATE-----```

It's not a valid PEM file and the ZLint command provides a fatal error. It does not panic. This is pretty reasonable. It wasn't presented with a valid certificate to lint.

When you do pass in a valid PEM, 
```-----BEGIN CERTIFICATE-----
MIIEyTCCA7GgAwIBAgINBnUDUrQShFEl9jfPmDANBgkqhkiG9w0BAQsFADBBMRUwEwYDVQQDDAxUZXN0IFN1YkNBIDIxDTALBgNVBAsMBFRlc3QxDDAKBgNVBAoMA01URzELMAkGA1UEBhMCREUwHhcNMTkxMTA0MDYzNTM2WhcNMjAxMTA0MDYzNTM2WjCBojEMMAoGA1UECgwDTVRHMRIwEAYDVQQHDAlEYXJtc3RhZHQxDzANBgNVBAgMBkhlc3NlbjELMAkGA1UEBhMCREUxHTAbBgNVBA8MFFByaXZhdGUgT3JnYW5pemF0aW9uMRgwFgYDVQRhGw9WQVRERS0xMjM0NTY3ODkxEjAQBgNVBAUTCTEyMzQ1Njc4OTETMBEGCysGAQQBgjc8AgEDDAJERTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANjJFjxxhlF2HKfm9gFWp4DMwUS5Qhb9Jomq2O425As5PCe9VCFNmV1X8a6wmADuWPPH3UD0BleveWnbtsj1IpU1Eo6uPwqeZQGJIKRuz6wgx9AZcbpXT5zgmf5VDDoVFn7sYvSc4a65FkfE2Ontamf0YuMDfpbZRS7nBzfojQFzKiB0OGvyC4icBMwGm+y11qwpuOImTzRljj3rAEboazF3o+8S8TyY712aXPQno4c3MSM1EqtZO+ILSBUy0ULzVWMFybI/WsnlqkpIpHQSFDX7V5ck8uiLLPrb/WnuE9EQXj4c2B8m8WlpUlAP9fMtqeddOqWSnxx4FAxhyUp9DPUCAwEAAaOCAVwwggFYMB8GA1UdIwQYMBaAFJZgP5yaf4gl4AHNZwhX0IYm8PFKMB0GA1UdDgQWBBTgmXRqGmcO54swL6B3XM95sawpYTAOBgNVHQ8BAf8EBAMCBaAwDAYDVR0TAQH/BAIwADAaBgNVHREEEzARgg93d3cuZXhhbXBsZS5jb20wYgYIKwYBBQUHAQEEVjBUMCgGCCsGAQUFBzAChhxodHRwOi8vY2EuZXhhbXBsZS5jb20vY2EuY3J0MCgGCCsGAQUFBzABhhxodHRwOi8vb2NzcC5leGFtcGxlLmNvbS9vY3NwMFkGA1UdIARSMFAwBwYFZ4EMAQEwRQYKKwYBBAG9Rw0YATA3MDUGCCsGAQUFBwIBFilodHRwOi8vd3d3LnRlbGVzZWMuZGUvc2VydmVycGFzcy9jcHMuaHRtbDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDQYJKoZIhvcNAQELBQADggEBAFRO6OOXlSsTAKv4Hp2jvIzMQIwwTZJMkHs4qshiFVOrLSXe3dDxxBjgn87rbXL0X6rqHysnSzhlhbW6bUg2gAAcyXf4DBu/TEwYUhlLkzSFFHM9tioHWZmL1sgedl2x4vwnZbMpqRlQTINwWbV0Ey5TfqGw84LoC0SXCzRoV3ta1tsZDIXS5UBuEZXMJ2ElvjqYbEnJhHjUqWRuCtbVlRHJ05EqigDLzv2Qm0Q5MluQq/Qc4SK1WZQ+PGOv2yU/VXdItYirD4LYBAsI2o2Abq5o/tjtjQeynj3NxqrF1Qd0/SHjQXE+ZMU+M7glP9e3jjwd9liavUKsF8O6wLha73g=
-----END CERTIFICATE-----```

The library does not panic and produces results as expected:

```{"e_basic_constraints_not_critical":{"result":"NA"},"e_ca_common_name_missing":{"result":"NA"},"e_ca_country_name_invalid":{"result":"NA"},"e_ca_country_name_missing":{"result":"NA"},"e_ca_crl_sign_not_set":{"result":"NA"},"e_ca_is_ca":{"result":"NA"},"e_ca_key_cert_sign_not_set":{"result":"NA"},"e_ca_key_usage_missing":{"result":"NA"},"e_ca_key_usage_not_critical":{"result":"NA"},"e_ca_organization_name_missing":{"result":"NA"},"e_ca_subject_field_empty":{"result":"NA"},"e_cab_dv_conflicts_with_locality":{"result":"NA"},"e_cab_dv_conflicts_with_org":{"result":"NA"},"e_cab_dv_conflicts_with_postal":{"result":"NA"},"e_cab_dv_conflicts_with_province":{"result":"NA"},"e_cab_dv_conflicts_with_street":{"result":"NA"},"e_cab_iv_requires_personal_name":{"result":"NA"},"e_cab_ov_requires_org":{"result":"NA"},"e_cert_contains_unique_identifier":{"result":"pass"},"e_cert_extensions_version_not_3":{"result":"pass"},"e_cert_policy_iv_requires_country":{"result":"NA"},"e_cert_policy_iv_requires_province_or_locality":{"result":"NA"},"e_cert_policy_ov_requires_country":{"result":"NA"},"e_cert_policy_ov_requires_province_or_locality":{"result":"NA"},"e_cert_unique_identifier_version_not_2_or_3":{"result":"NA"},"e_distribution_point_incomplete":{"result":"NA"},"e_dnsname_bad_character_in_label":{"result":"pass"},"e_dnsname_contains_bare_iana_suffix":{"result":"pass"},"e_dnsname_empty_label":{"result":"pass"},"e_dnsname_hyphen_in_sld":{"result":"pass"},"e_dnsname_label_too_long":{"result":"pass"},"e_dnsname_left_label_wildcard_correct":{"result":"pass"},"e_dnsname_not_valid_tld":{"result":"pass"},"e_dnsname_underscore_in_sld":{"result":"pass"},"e_dnsname_wildcard_only_in_left_label":{"result":"pass"},"e_dsa_correct_order_in_subgroup":{"result":"NA"},"e_dsa_improper_modulus_or_divisor_size":{"result":"NA"},"e_dsa_params_missing":{"result":"NA"},"e_dsa_shorter_than_2048_bits":{"result":"NA"},"e_dsa_unique_correct_representation":{"result":"NA"},"e_ec_improper_curves":{"result":"NA"},"e_ev_business_category_missing":{"result":"pass"},"e_ev_country_name_missing":{"result":"pass"},"e_ev_organization_name_missing":{"result":"pass"},"e_ev_serial_number_missing":{"result":"pass"},"e_ev_valid_time_too_long":{"result":"pass"},"e_ext_aia_marked_critical":{"result":"pass"},"e_ext_authority_key_identifier_critical":{"result":"pass"},"e_ext_authority_key_identifier_missing":{"result":"pass"},"e_ext_authority_key_identifier_no_key_identifier":{"result":"pass"},"e_ext_cert_policy_disallowed_any_policy_qualifier":{"result":"pass"},"e_ext_cert_policy_duplicate":{"result":"pass"},"e_ext_cert_policy_explicit_text_ia5_string":{"result":"NA"},"e_ext_cert_policy_explicit_text_too_long":{"result":"NA"},"e_ext_duplicate_extension":{"result":"pass"},"e_ext_freshest_crl_marked_critical":{"result":"NA"},"e_ext_ian_dns_not_ia5_string":{"result":"NA"},"e_ext_ian_empty_name":{"result":"NA"},"e_ext_ian_no_entries":{"result":"NA"},"e_ext_ian_rfc822_format_invalid":{"result":"NA"},"e_ext_ian_space_dns_name":{"result":"NA"},"e_ext_ian_uri_format_invalid":{"result":"NA"},"e_ext_ian_uri_host_not_fqdn_or_ip":{"result":"NA"},"e_ext_ian_uri_not_ia5":{"result":"NA"},"e_ext_ian_uri_relative":{"result":"NA"},"e_ext_key_usage_cert_sign_without_ca":{"result":"pass"},"e_ext_key_usage_without_bits":{"result":"pass"},"e_ext_name_constraints_not_critical":{"result":"NA"},"e_ext_name_constraints_not_in_ca":{"result":"NA"},"e_ext_policy_constraints_empty":{"result":"NA"},"e_ext_policy_constraints_not_critical":{"result":"NA"},"e_ext_policy_map_any_policy":{"result":"NA"},"e_ext_san_contains_reserved_ip":{"result":"pass"},"e_ext_san_directory_name_present":{"result":"pass"},"e_ext_san_dns_name_too_long":{"result":"pass"},"e_ext_san_dns_not_ia5_string":{"result":"pass"},"e_ext_san_edi_party_name_present":{"result":"pass"},"e_ext_san_empty_name":{"result":"pass"},"e_ext_san_missing":{"result":"pass"},"e_ext_san_no_entries":{"result":"pass"},"e_ext_san_not_critical_without_subject":{"result":"pass"},"e_ext_san_other_name_present":{"result":"pass"},"e_ext_san_registered_id_present":{"result":"pass"},"e_ext_san_rfc822_format_invalid":{"result":"pass"},"e_ext_san_rfc822_name_present":{"result":"pass"},"e_ext_san_space_dns_name":{"result":"pass"},"e_ext_san_uniform_resource_identifier_present":{"result":"pass"},"e_ext_san_uri_format_invalid":{"result":"pass"},"e_ext_san_uri_host_not_fqdn_or_ip":{"result":"pass"},"e_ext_san_uri_not_ia5":{"result":"pass"},"e_ext_san_uri_relative":{"result":"pass"},"e_ext_subject_directory_attr_critical":{"result":"NA"},"e_ext_subject_key_identifier_critical":{"result":"pass"},"e_ext_subject_key_identifier_missing_ca":{"result":"NA"},"e_ext_tor_service_descriptor_hash_invalid":{"result":"NA"},"e_generalized_time_does_not_include_seconds":{"result":"NA"},"e_generalized_time_includes_fraction_seconds":{"result":"NA"},"e_generalized_time_not_in_zulu":{"result":"NA"},"e_ian_bare_wildcard":{"result":"NA"},"e_ian_dns_name_includes_null_char":{"result":"NA"},"e_ian_dns_name_starts_with_period":{"result":"NA"},"e_ian_wildcard_not_first":{"result":"NA"},"e_inhibit_any_policy_not_critical":{"result":"NA"},"e_international_dns_name_not_nfc":{"result":"pass"},"e_international_dns_name_not_unicode":{"result":"pass"},"e_invalid_certificate_version":{"result":"pass"},"e_issuer_dn_country_not_printable_string":{"result":"pass"},"e_issuer_field_empty":{"result":"pass"},"e_name_constraint_empty":{"result":"NA"},"e_name_constraint_maximum_not_absent":{"result":"NA"},"e_name_constraint_minimum_non_zero":{"result":"NA"},"e_old_root_ca_rsa_mod_less_than_2048_bits":{"result":"NA"},"e_old_sub_ca_rsa_mod_less_than_1024_bits":{"result":"NA"},"e_old_sub_cert_rsa_mod_less_than_1024_bits":{"result":"NA"},"e_onion_subject_validity_time_too_large":{"result":"NA"},"e_path_len_constraint_improperly_included":{"result":"pass"},"e_path_len_constraint_zero_or_less":{"result":"pass"},"e_public_key_type_not_allowed":{"result":"pass"},"e_qcstatem_etsi_present_qcs_critical":{"result":"NA"},"e_qcstatem_etsi_type_as_statem":{"result":"NA"},"e_qcstatem_mandatory_etsi_statems":{"result":"NA"},"e_qcstatem_qccompliance_valid":{"result":"NA"},"e_qcstatem_qclimitvalue_valid":{"result":"NA"},"e_qcstatem_qcpds_valid":{"result":"NA"},"e_qcstatem_qcretentionperiod_valid":{"result":"NA"},"e_qcstatem_qcsscd_valid":{"result":"NA"},"e_qcstatem_qctype_valid":{"result":"NA"},"e_root_ca_extended_key_usage_present":{"result":"NA"},"e_root_ca_key_usage_must_be_critical":{"result":"NA"},"e_root_ca_key_usage_present":{"result":"NA"},"e_rsa_exp_negative":{"result":"pass"},"e_rsa_mod_less_than_2048_bits":{"result":"pass"},"e_rsa_no_public_key":{"result":"pass"},"e_rsa_public_exponent_not_odd":{"result":"pass"},"e_rsa_public_exponent_too_small":{"result":"pass"},"e_san_bare_wildcard":{"result":"pass"},"e_san_dns_name_includes_null_char":{"result":"pass"},"e_san_dns_name_onion_not_ev_cert":{"result":"NA"},"e_san_dns_name_starts_with_period":{"result":"pass"},"e_san_wildcard_not_first":{"result":"pass"},"e_serial_number_longer_than_20_octets":{"result":"pass"},"e_serial_number_not_positive":{"result":"pass"},"e_signature_algorithm_not_supported":{"result":"pass"},"e_spki_rsa_encryption_parameter_not_null":{"result":"pass"},"e_sub_ca_aia_does_not_contain_ocsp_url":{"result":"NA"},"e_sub_ca_aia_marked_critical":{"result":"NA"},"e_sub_ca_aia_missing":{"result":"NA"},"e_sub_ca_certificate_policies_missing":{"result":"NA"},"e_sub_ca_crl_distribution_points_does_not_contain_url":{"result":"NA"},"e_sub_ca_crl_distribution_points_marked_critical":{"result":"NA"},"e_sub_ca_crl_distribution_points_missing":{"result":"NA"},"e_sub_cert_aia_does_not_contain_ocsp_url":{"result":"pass"},"e_sub_cert_aia_marked_critical":{"result":"pass"},"e_sub_cert_aia_missing":{"result":"pass"},"e_sub_cert_cert_policy_empty":{"result":"pass"},"e_sub_cert_certificate_policies_missing":{"result":"pass"},"e_sub_cert_country_name_must_appear":{"result":"pass"},"e_sub_cert_crl_distribution_points_does_not_contain_url":{"result":"NA"},"e_sub_cert_crl_distribution_points_marked_critical":{"result":"NA"},"e_sub_cert_eku_missing":{"result":"pass"},"e_sub_cert_eku_server_auth_client_auth_missing":{"result":"pass"},"e_sub_cert_given_name_surname_contains_correct_policy":{"result":"NA"},"e_sub_cert_key_usage_cert_sign_bit_set":{"result":"pass"},"e_sub_cert_key_usage_crl_sign_bit_set":{"result":"pass"},"e_sub_cert_locality_name_must_appear":{"result":"pass"},"e_sub_cert_locality_name_must_not_appear":{"result":"pass"},"e_sub_cert_not_is_ca":{"result":"pass"},"e_sub_cert_or_sub_ca_using_sha1":{"result":"pass"},"e_sub_cert_postal_code_must_not_appear":{"result":"pass"},"e_sub_cert_province_must_appear":{"result":"pass"},"e_sub_cert_province_must_not_appear":{"result":"pass"},"e_sub_cert_street_address_should_not_exist":{"result":"pass"},"e_sub_cert_valid_time_longer_than_39_months":{"result":"pass"},"e_sub_cert_valid_time_longer_than_825_days":{"result":"pass"},"e_subject_common_name_max_length":{"result":"pass"},"e_subject_common_name_not_from_san":{"result":"NA"},"e_subject_contains_noninformational_value":{"result":"pass"},"e_subject_contains_reserved_arpa_ip":{"result":"NA"},"e_subject_contains_reserved_ip":{"result":"pass"},"e_subject_country_not_iso":{"result":"pass"},"e_subject_dn_country_not_printable_string":{"result":"pass"},"e_subject_dn_not_printable_characters":{"result":"pass"},"e_subject_dn_serial_number_max_length":{"result":"pass"},"e_subject_dn_serial_number_not_printable_string":{"result":"pass"},"e_subject_email_max_length":{"result":"pass"},"e_subject_empty_without_san":{"result":"pass"},"e_subject_given_name_max_length":{"result":"pass"},"e_subject_info_access_marked_critical":{"result":"NA"},"e_subject_locality_name_max_length":{"result":"pass"},"e_subject_not_dn":{"result":"pass"},"e_subject_organization_name_max_length":{"result":"pass"},"e_subject_organizational_unit_name_max_length":{"result":"pass"},"e_subject_postal_code_max_length":{"result":"pass"},"e_subject_printable_string_badalpha":{"result":"pass"},"e_subject_state_name_max_length":{"result":"pass"},"e_subject_street_address_max_length":{"result":"pass"},"e_subject_surname_max_length":{"result":"pass"},"e_tbs_signature_rsa_encryption_parameter_not_null":{"result":"pass"},"e_utc_time_does_not_include_seconds":{"result":"pass"},"e_utc_time_not_in_zulu":{"result":"pass"},"e_validity_time_not_positive":{"result":"pass"},"e_wrong_time_format_pre2050":{"result":"pass"},"n_ca_digital_signature_not_set":{"result":"NA"},"n_contains_redacted_dnsname":{"result":"pass"},"n_ecdsa_ee_invalid_ku":{"result":"NA"},"n_multiple_subject_rdn":{"result":"pass"},"n_san_dns_name_duplicate":{"result":"pass"},"n_sub_ca_eku_missing":{"result":"NA"},"n_sub_ca_eku_not_technically_constrained":{"result":"NA"},"n_subject_common_name_included":{"result":"pass"},"w_ct_sct_policy_count_unsatisfied":{"result":"info","details":"Certificate had 0 embedded SCTs. Browser policy may require 2 for this certificate."},"w_distribution_point_missing_ldap_or_uri":{"result":"NA"},"w_dnsname_underscore_in_trd":{"result":"pass"},"w_dnsname_wildcard_left_of_public_suffix":{"result":"pass"},"w_eku_critical_improperly":{"result":"pass"},"w_ext_aia_access_location_missing":{"result":"pass"},"w_ext_cert_policy_contains_noticeref":{"result":"pass"},"w_ext_cert_policy_explicit_text_includes_control":{"result":"NA"},"w_ext_cert_policy_explicit_text_not_nfc":{"result":"NA"},"w_ext_cert_policy_explicit_text_not_utf8":{"result":"NA"},"w_ext_crl_distribution_marked_critical":{"result":"NA"},"w_ext_ian_critical":{"result":"NA"},"w_ext_key_usage_not_critical":{"result":"pass"},"w_ext_policy_map_not_critical":{"result":"NA"},"w_ext_policy_map_not_in_cert_policy":{"result":"NA"},"w_ext_san_critical_with_subject_dn":{"result":"pass"},"w_ext_subject_key_identifier_missing_sub_cert":{"result":"pass"},"w_ian_iana_pub_suffix_empty":{"result":"NA"},"w_issuer_dn_leading_whitespace":{"result":"pass"},"w_issuer_dn_trailing_whitespace":{"result":"pass"},"w_multiple_issuer_rdn":{"result":"pass"},"w_name_constraint_on_edi_party_name":{"result":"NA"},"w_name_constraint_on_registered_id":{"result":"NA"},"w_name_constraint_on_x400":{"result":"NA"},"w_qcstatem_qcpds_lang_case":{"result":"NA"},"w_qcstatem_qctype_web":{"result":"NA"},"w_root_ca_basic_constraints_path_len_constraint_field_present":{"result":"NA"},"w_root_ca_contains_cert_policy":{"result":"NA"},"w_rsa_mod_factors_smaller_than_752":{"result":"pass"},"w_rsa_mod_not_odd":{"result":"pass"},"w_rsa_public_exponent_not_in_range":{"result":"pass"},"w_san_iana_pub_suffix_empty":{"result":"pass"},"w_sub_ca_aia_does_not_contain_issuing_ca_url":{"result":"NA"},"w_sub_ca_certificate_policies_marked_critical":{"result":"NA"},"w_sub_ca_eku_critical":{"result":"NA"},"w_sub_ca_name_constraints_not_critical":{"result":"NA"},"w_sub_cert_aia_does_not_contain_issuing_ca_url":{"result":"pass"},"w_sub_cert_certificate_policies_marked_critical":{"result":"pass"},"w_sub_cert_eku_extra_values":{"result":"pass"},"w_sub_cert_sha1_expiration_too_long":{"result":"NA"},"w_subject_contains_malformed_arpa_ip":{"result":"NA"},"w_subject_dn_leading_whitespace":{"result":"pass"},"w_subject_dn_trailing_whitespace":{"result":"pass"}}
```

As such this doesn't look like a bug.

**Comment by cryptosource-GmbH** [comment]
What I meant was that the in the unit tests the panic is uncaught. I simply added the cert to the list of test certificates in one of the test files. When running 
    go test -v lints/*
this results in 
```go
--- FAIL: TestEtsiQcCompliance (0.00s)
panic: PEM decode failed! [recovered]
        panic: PEM decode failed!

goroutine 431 [running]:
testing.tRunner.func1(0xc0000f4200)
        /usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x7cbb80, 0x90d4e0)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
command-line-arguments.ReadCertificate(0xc00029afe0, 0x1c, 0x16)
        /home/fstrenzke/go/src/github.com/zmap/zlint/lints/testingUtil.go:40 +0x1fa
command-line-arguments.TestEtsiQcCompliance(0xc0000f4200)
        /home/fstrenzke/go/src/github.com/zmap/zlint/lints/lint_qcstatem_qccompliance_valid_test.go:35 +0x357
testing.tRunner(0xc0000f4200, 0x8b1000)
        /usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x350
FAIL    command-line-arguments  0.178s
FAIL
```

But I assume the only way to deal with this would be to recover from the panic within the lint and possibly then return a fatal result.

**Comment by cpu** [comment]
@cryptosource-GmbH (_are you also @falko-strenzke btw or are we talking to two people?_) I agree with @zakird's assessment here. The panic in your last comment seems like working-as-expected behaviour to me. If you add a corrupt test file to a unit test a panic seems like a reasonable response. I see it as a developer facing problem to a developer introduced issue. The `ReadCertificate` function from `lints/testingUtils.go` in your trace was deliberately implemented to panic on invalid test input:

https://github.com/zmap/zlint/blob/7314deb0a2a11b829a1144d21ec75d16b587e8f4/lints/testingUtil.go#L38-L41

> I assume the only way to deal with this would be to recover from the panic within the lint and possibly then return a fatal result.

I think a more targeted fix would be to change the implementation of `ReadCertificate` to return `(*x509.Certificate, error)`, return `nil` and an error when there's a PEM decode panic, and then update all of the callers to check for a non-nil error.

I'm mostly sharing the above to point out an alternative. I don't think it is a good idea to implement. It's a fair amount of find/replace work for the small win of changing the developer experience of adding buggy test data from seeing a panic to seeing an error.

**Comment by cpu** [comment]
I'm going to close this issue since it seems we've settled on a wontfix outcome. The discussion can continue in this thread if needed. Thanks!

**Comment by cryptosource-GmbH** [comment]
yes, it's the same person ;-) I agree, we can close the thread. Thank you both for your suggestions.
